scripts/pythondistdeps.py

```diff
@@ -11,21 +11,77 @@
 # RPM python dependency generator, using .egg-info/.egg-link/.dist-info data
 #
 
-# Please know:
-# - Notes from an attempted rewrite from pkg_resources to importlib.metadata in
-#   2020 can be found in the message of the commit that added this line.
-
 from __future__ import print_function
 import argparse
-from os.path import basename, dirname, isdir, sep
-from sys import argv, stdin, version
 from distutils.sysconfig import get_python_lib
+from itertools import chain
+from os.path import dirname, sep
+from packaging.requirements import Requirement
+from packaging.version import parse
+import re
+from sys import argv, stdin
 from warnings import warn
 
+try:
+    from importlib.metadata import PathDistribution
+except ImportError:
+    from importlib_metadata import PathDistribution
+
+try:
+    from pathlib import Path
+except ImportError:
+    from pathlib2 import Path
+
+
+class Req(Requirement):
+    @property
+    def key(self):
+        return self.name.lower().replace('_', '-')
```

Comment:
@torsava This is the implementation of key. I don't believe it matches the original key value in all cases.

Reply:
Oh, how come? You think the [`Requirement.key` property](https://setuptools.readthedocs.io/en/latest/pkg_resources.html) gave different results in different situations?

Reply:
To clarify a possible confusion, I think `key` always returned the same thing. But this script had two modes, in legacy mode it used the `key` value, and in pep503 mode it used the `normalize_name` function.

So yes, there are two different methods to convert the name, but that is handled by this script, not by `Requirement.key`.

Reply:
That really depends on `self.name` value. I don't see that from the code here.

In pkg_resources, key is based on `self.project_name` and that [is set to](https://github.com/pypa/setuptools/blob/73e379cc55ac1e9ec63c4ac30b75ecc82418f513/pkg_resources/__init__.py#L2564) `safe_name(project_name)`.

In packaging, I believe `self.name` is not pre-processed. See:

```pycon
>>> import pkg_resources
>>> pkg_resources.safe_name("Tree-_-FALLING.Down").lower()
'tree-falling.down'
>>> from packaging.requirements import Requirement
>>> Requirement("Tree-_-FALLING.Down").name.lower().replace('_', '-')
'tree---falling.down'
```

Reply:
Ah, I see. Yeah, that should be fixed as well. @s-t-e-v-e-n-k please take a look.

Reply:
Instead, I'd create two properties here:

```python
class Req(Requirement):
    @property
    def legacy_normalized_name(self):
        """Like pkg_resources key property"""
        if not self._legacy_normalized_name:
            self._legacy_normalized_name = re.sub(r'[-_]+', '-', self.name).lower()
￼       return self._legacy_normalized_name

    @property
￼   def normalized_name(self):
        """PEP 503 normalized name"""
        if not self._normalized_name:
            self._normalized_name = re.sub(r'[-_.]+', '-', self.name).lower()
￼       return self._normalized_name
```

"key" is not very descriptive.

Reply:
Which is a massive layering violation. I'll have a think on this.

Reply:
massive layering violation?

Reply:
I'm also confused as to what that means..?

Reply:
I'm sorry, I misread that completely and conflated it with the argument parsing, my apologies.

This would also require massive changes everywhere to stop using the key property, which @torsava has seemed against -- but I'm happy to do so.

Secondly, this doesn't solve it for Distribution, so anything we do for requirements we should also do for the Distribution class -- my current idea is a classmethod on Req that we call there and in Distribution. And a consequence the `normalize_name()` function gets removed as well.

`re.sub('[^A-Za-z0-9.]+', '-', self.name).lower()` is what Distribution current does, which is what `pkg_resources.safe_name()` does.

Reply:
@s-t-e-v-e-n-k I wasn't against ceasing to use the `key` property, I was against passing different values in it's place. The name of the property/function is immaterial.

Reply:
And I agree with unifying the two ways to get a normalized name.